### PR TITLE
haskell-generic-builder: pass pname and version as is

### PR DESCRIPTION
**Pull request by sternenseemann**
This is a full set rebuild, however it improves the name generation
for the static and cross case since the respective additional
components are now inserted between pname and version instead of after
name like before. This prevents builtins.parseDrvName from mistaking a
platform config string for a version component.

(Untested so far)

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sternenseemann** [comment]
Built `/nix/store/lvs6rvvnlkid6g0wv7qxsr62nfndpvxc-cabal2nix-unstable-2021-06-12` to verify it's good.

**Review by maralorn — APPROVED** [review]
LGTM. Will merge this after the haskell-updates merge.

**Comment by cdepillabout** [comment]
@sternenseemann When doing a static or cross build, what is an example of what a package `name` looked like before this change?  What does it look like now with this change?

**Comment by sternenseemann** [comment]
```console
$ nix-instantiate -A pkgsStatic.haskellPackages.hello -A pkgsCross.aarch64-multiplatform.haskellPackages.hello
/nix/store/0nvxybp8av7gczd939hp8bh71x5f0vv2-hello-static-x86_64-unknown-linux-musl-1.0.0.2.drv
/nix/store/kj6cph9qwxigf446q5r8pcrrf1yrwscv-hello-aarch64-unknown-linux-gnu-1.0.0.2.drv
```

**Comment by cdepillabout** [comment]
@sternenseemann Ah, nice.  This looks like a strict improvement!

**Comment by sternenseemann** [comment]
See also <https://github.com/NixOS/nixpkgs/pull/120762> for additional context.

**Pull request merged**
